libyul/optimiser/README.md

```diff
@@ -82,22 +165,144 @@ function calls can only appear in the right-hand side of a variable declaration,
+inside the for loop.
+
+After this stage, the Redundant Assign Eliminator is recommended to remove the unnecessary
+intermediate assignments.
+
+This stage provides best results if the Expression Splitter and the Common Subexpression Eliminator
+is run right before it, because then it does not generate excessive amounts of variables.
+On the other hand, the Common Subexpression Eliminator could be more efficient if run after the
+SSA transform.
+
+### Redundant Assign Eliminator
+
+The SSA transform always generates an assignment of the form ``a := a_i``, even though
+these might be unnecessary in many cases, like the following example:
+
+    {
+        let a := 1
+        a := mload(a)
+        a := sload(a)
+        sstore(a, 1)
+    }
+
+The SSA transform converts this snippet to the following:
+
+    {
+        let a_1 := 1
+        a := a_1
+        let a_2 := mload(a_1)
+        a := a_2
+        let a_3 := sload(a_2)
+        a := a_3
+        sstore(a_3, 1)
+    }
+
+The Redundant Assign Eliminator removes all the three assignments to ``a``, because
+the value of ``a`` is not used and thus turn this
+snippet into strict SSA form:
+
+    {
+        let a_1 := 1
+        let a_2 := mload(a_1)
+        let a_3 := sload(a_2)
+        sstore(a_3, 1)
+    }
+
+Of course the intricate parts of determining whether an assignment is redundant or not
+are connected to joining control flow.
+
+The component works as follows in detail:
 
-## Common Subexpression Eliminator
+The AST is traversed twice: in an information gathering step and in the
+actual removal step. During information gathering, we maintain a
+mapping from assignment statements to the three states
+"unused", "undecided" and "used" which signifies whether the assigned
+value will be used later by a reference to the variable.
+
+When an assignment is visited, it is added to the mapping in the "undecided" state
+(see remark about for loops below) and every other assignment to the same variable
+that is still in the "undecided" state is changed to "unused".
+When a variable is referenced, the state of any assignment to that variable still
+in the "undecided" state is changed to "used".
+
+At points where control flow splits, a copy
+of the mapping is handed over to each branch. At points where control flow
+joins, the two mappings coming from the two branches are combined in the following way:
+Statements that are only in one mapping or have the same state are used unchanged.
+Conflicting values are resolved in the following way:
+
+ - "unused", "undecided" -> "undecided"
+ - "unused", "used" -> "used"
+ - "undecided, "used" -> "used"
+
+For for-loops, the condition, body and post-part are visited twice, taking
+the joining control-flow at the condition into account.
+In other words, we create three control flow paths: Zero runs of the loop,
+one run and two runs and then combine them at the end.
+Running at most twice is enough because there are only three different states.
```

Comment:
It would be nice to have a proof for this bold statement...

Reply:
"Since there are only three different states, two runs of the loop are clearly enough for convergence." q.e.d.